js/compiler/compiler.go

```diff
@@ -44,103 +44,119 @@ var (
-}
-
-// Constructs a new compiler.
-func New() (*Compiler, error) {
-	var err error
-	once.Do(func() {
-		compilerInstance, err = new()
-	})
-
-	return compilerInstance, err
-}
-
-func new() (*Compiler, error) {
-	conf := rice.Config{
-		LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
-	}
-
-	babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
-
-	c := &Compiler{vm: goja.New()}
-	if _, err := c.vm.RunString(babelSrc); err != nil {
-		return nil, err
-	}
+const (
+	// CompatibilityModeES6 is achieved with Babel and core.js
+	CompatibilityModeES6 CompatibilityMode = iota + 1
+	// CompatibilityModeES51 is standard goja
+	CompatibilityModeES51
+)
 
-	c.this = c.vm.Get("Babel")
-	thisObj := c.this.ToObject(c.vm)
-	if err := c.vm.ExportTo(thisObj.Get("transform"), &c.transform); err != nil {
-		return nil, err
-	}
+// A Compiler compiles JavaScript source code (ES5.1 or ES6) into a goja.Program
+type Compiler struct{}
 
-	return c, nil
+// New returns a new Compiler
+func New() *Compiler {
+	return &Compiler{}
 }
 
-// Transform the given code into ES5.
+// Transform the given code into ES5, while synchronizing to ensure only a single
+// babel instance is in use.
 func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
-	opts := make(map[string]interface{})
-	for k, v := range DefaultOpts {
-		opts[k] = v
-	}
-	opts["filename"] = filename
-
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-
-	startTime := time.Now()
-	v, err := c.transform(c.this, c.vm.ToValue(src), c.vm.ToValue(opts))
-	if err != nil {
-		return code, srcmap, err
-	}
-	logrus.WithField("t", time.Since(startTime)).Debug("Babel: Transformed")
-	vO := v.ToObject(c.vm)
-
-	if err := c.vm.ExportTo(vO.Get("code"), &code); err != nil {
-		return code, srcmap, err
-	}
-
-	var rawmap map[string]interface{}
-	if err := c.vm.ExportTo(vO.Get("map"), &rawmap); err != nil {
-		return code, srcmap, err
-	}
-	if err := mapstructure.Decode(rawmap, &srcmap); err != nil {
-		return code, srcmap, err
+	var b *babel
+	if b, err = newBabel(); err != nil {
+		return
 	}
 
-	return code, srcmap, nil
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	return b.Transform(src, filename)
 }
 
-// Compiles the program, first trying ES5, then ES6.
-func (c *Compiler) Compile(src, filename string, pre, post string, strict bool) (*goja.Program, string, error) {
-	return c.compile(src, filename, pre, post, strict, true)
+// Compile the program
+func (c *Compiler) Compile(src, filename string, pre, post string,
+	strict bool, compatMode CompatibilityMode) (*goja.Program, string, error) {
+	return c.compile(src, filename, pre, post, strict, compatMode)
 }
 
-func (c *Compiler) compile(src, filename string, pre, post string, strict, tryBabel bool) (*goja.Program, string, error) {
+func (c *Compiler) compile(src, filename string, pre, post string,
+	strict bool, compatMode CompatibilityMode) (*goja.Program, string, error) {
```

Comment:
Now that the signatures are the same, we can just expose this as the public `Compile`, right?